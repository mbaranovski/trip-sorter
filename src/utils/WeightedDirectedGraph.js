export default class WeightedDirectedGraph {
  graph = {};

  contains(node) {
    return !!this.graph[node];
  }

  containsEdge(startNode, endNode) {
    return !!this.graph[startNode].edges[endNode];
  }

  addVertex(node) {
    if (!this.contains(node)) {
      this.graph[node] = { edges: {} };
    }
  }

  removeVertex(node) {
    if (this.contains(node)) {
      for (let connectedNode in this.graph[node].edges) {
        if (this.graph[node].edges.hasOwnProperty(connectedNode))
          this.removeEdge(node, connectedNode);
      }
      delete this.graph[node];
    }
  }

  addEdge = (startNode, endNode, weight, rest) => {
    if (this.contains(startNode) && this.contains(endNode)) {
      this.graph[startNode].edges[endNode] = { weight, ...rest };
    }
  };

  removeEdge = (startNode, endNode) => {
    if (this.contains(startNode) && this.contains(endNode)) {
      delete this.graph[startNode].edges[endNode];
    }
  };
}
