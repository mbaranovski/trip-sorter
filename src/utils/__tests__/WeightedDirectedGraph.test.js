import WeightedDirectedGraph from "../WeightedDirectedGraph";

test("Creates an instance of WeightedDirectedGraph and adds/removes simple vertices and edges", () => {
  const graph = new WeightedDirectedGraph();

  graph.addVertex("A");
  graph.addVertex("B");
  graph.addVertex("C");

  graph.addEdge("A", "B", 1, { ref: "AB1" });
  graph.addEdge("B", "A", 2, { ref: "BA1" });
  graph.addEdge("B", "C", 5, { ref: "BC1" });

  expect(graph.graph).toMatchSnapshot();

  graph.removeEdge("B", "A");
  graph.removeVertex("B");

  expect(graph.graph).toMatchSnapshot();
});
