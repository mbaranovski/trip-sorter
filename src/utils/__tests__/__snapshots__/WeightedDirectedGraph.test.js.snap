// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creates an instance of WeightedDirectedGraph and adds/removes simple vertices and edges 1`] = `
Object {
  "A": Object {
    "edges": Object {
      "B": Object {
        "ref": "AB1",
        "weight": 1,
      },
    },
  },
  "B": Object {
    "edges": Object {
      "A": Object {
        "ref": "BA1",
        "weight": 2,
      },
      "C": Object {
        "ref": "BC1",
        "weight": 5,
      },
    },
  },
  "C": Object {
    "edges": Object {},
  },
}
`;

exports[`Creates an instance of WeightedDirectedGraph and adds/removes simple vertices and edges 2`] = `
Object {
  "A": Object {
    "edges": Object {
      "B": Object {
        "ref": "AB1",
        "weight": 1,
      },
    },
  },
  "C": Object {
    "edges": Object {},
  },
}
`;
